File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/release/ReleasePresenterImpl.java
package com.twtstudio.tjwhm.lostfound.release;

import com.twt.wepeiyang.commons.network.RetrofitProvider;
import com.twt.wepeiyang.commons.network.RxErrorHandler;
import com.twtstudio.tjwhm.lostfound.base.BaseBean;
import com.twtstudio.tjwhm.lostfound.base.CallbackBean;

import java.io.File;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by tjwhm on 2017/7/6 15:55.
 * Happy coding!
 **/

public class ReleasePresenterImpl implements ReleaseContract.ReleasePresenter {

    ReleaseContract.ReleaseView releaseView;
    ReleaseApi releaseApi;

    public ReleasePresenterImpl(ReleaseContract.ReleaseView releaseView) {
        this.releaseView = releaseView;
    }


    @Override
    public void uploadReleaseData(final Map<String, Object> map, String lostOrFound) {
        releaseApi = RetrofitProvider.getRetrofit().create(ReleaseApi.class);
        releaseApi.updateRelease(map, lostOrFound)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::successCallBack, new RxErrorHandler());
    }

    @Override
    public void uploadReleaseDataWithPic(Map<String, Object> map, String lostOrFound, File file) {
        releaseApi = RetrofitProvider.getRetrofit().create(ReleaseApi.class);

        MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);//表单类型
        RequestBody imageBody = RequestBody.create(MediaType.parse("multipart/form-data"), file);
        builder.addFormDataPart("pic[]", file.getName(), imageBody);//imgfile 后台接收图片流的参数名
        builder.addFormDataPart("title", String.valueOf(map.get("title")));
        builder.addFormDataPart("time", String.valueOf(map.get("time")));
        builder.addFormDataPart("place", String.valueOf(map.get("place")));
        builder.addFormDataPart("detail_type", String.valueOf(map.get("detail_type")));
        builder.addFormDataPart("card_number", String.valueOf(map.get("card_number")));
        builder.addFormDataPart("card_name", String.valueOf(map.get("card_name")));
        builder.addFormDataPart("name", String.valueOf(map.get("name")));
        builder.addFormDataPart("phone", String.valueOf(map.get("phone")));
        builder.addFormDataPart("item_description", String.valueOf(map.get("item_description")));
        builder.addFormDataPart("other_tag", "");
        builder.addFormDataPart("duration", String.valueOf(map.get("duration")));
        List<MultipartBody.Part> parts = builder.build().parts();
        releaseApi.updateReleaseWithPic(lostOrFound, parts)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::successCallBack, new RxErrorHandler());
    }

    @Override
    public void successCallBack(BaseBean baseBean) {
        releaseView.successCallBack(baseBean);
    }


    @Override
    public void successEditCallback(BaseBean baseBean) {
        releaseView.successCallBack(baseBean);
    }

    @Override
    public void delete(int id) {
        releaseApi = RetrofitProvider.getRetrofit().create(ReleaseApi.class);
        releaseApi.delete(String.valueOf(id))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::deleteSuccessCallBack, new RxErrorHandler());
    }

    @Override
    public void deleteSuccessCallBack(BaseBean baseBean) {
        releaseView.deleteSuccessCallBack();
    }

    @Override
    public void uploadEditDataWithPic(Map<String, Object> map, String lostOrFound, File file, int id) {
        releaseApi = RetrofitProvider.getRetrofit().create(ReleaseApi.class);

        MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);
        if(file!=null) {//表单类型
            RequestBody imageBody = RequestBody.create(MediaType.parse("multipart/form-data"), file);
            builder.addFormDataPart("pic[]", file.getName(), imageBody);
        }
        //imgfile 后台接收图片流的参数名
        builder.addFormDataPart("title", String.valueOf(map.get("title")));
        builder.addFormDataPart("time", String.valueOf(map.get("time")));
        builder.addFormDataPart("place", String.valueOf(map.get("place")));
        builder.addFormDataPart("detail_type", String.valueOf(map.get("detail_type")));
        builder.addFormDataPart("card_number", String.valueOf(map.get("card_number")));
        builder.addFormDataPart("card_name", String.valueOf(map.get("card_name")));
        builder.addFormDataPart("name", String.valueOf(map.get("name")));
        builder.addFormDataPart("phone", String.valueOf(map.get("phone")));
        builder.addFormDataPart("item_description", String.valueOf(map.get("item_description")));
        builder.addFormDataPart("other_tag", "");
        builder.addFormDataPart("duration", String.valueOf(map.get("duration")));
        List<MultipartBody.Part> parts = builder.build().parts();
        String lof;
        if (Objects.equals(lostOrFound, "editLost")) {
            lof = "lost";
        } else {
            lof = "found";
        }
        releaseApi.updateEditWithPic(lof, String.valueOf(id), parts)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::successEditCallback, new RxErrorHandler());
    }


}

File: WePeiYang/lostfound2/src/main/java/com/example/lostfond2/release/ReleasePresenterImpl.kt
package com.example.lostfond2.release

import android.util.Log
import com.example.lostfond2.service.InverseID
import com.example.lostfond2.service.LostFoundService
import com.example.lostfond2.service.MyListDataOrSearchBean
import com.twt.wepeiyang.commons.experimental.extensions.QuietCoroutineExceptionHandler
import com.twt.wepeiyang.commons.experimental.extensions.awaitAndHandle
import com.twt.wepeiyang.commons.experimental.network.CommonBody
import com.twt.wepeiyang.commons.network.RetrofitProvider
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import org.jetbrains.anko.coroutines.experimental.bg
import java.io.File
import java.util.*
import kotlin.collections.ArrayList

class ReleasePresenterImpl(var releaseView: ReleaseContract.ReleaseView) : ReleaseContract.ReleasePresenter {


    override fun uploadReleaseData(map: Map<String, Any>, lostOrFound: String) {

        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        builder.apply {
            addFormDataPart("title", map["title"].toString())
            addFormDataPart("time", map["time"].toString())
            addFormDataPart("place", map["place"].toString())
            addFormDataPart("detail_type", map["detail_type"].toString())
            addFormDataPart("card_number", map["card_number"].toString())
            addFormDataPart("card_name", map["card_name"].toString())
            addFormDataPart("name", map["name"].toString())
            addFormDataPart("phone", map["phone"].toString())
            addFormDataPart("item_description", map["item_description"].toString())
            addFormDataPart("other_tag", "")
            addFormDataPart("duration", map["duration"].toString())
            addFormDataPart("QQ", map["QQ"].toString())
            addFormDataPart("Wechat", map["Wechat"].toString())
            addFormDataPart("recapture_place", map["recapture_place"].toString())
            addFormDataPart("recapture_enterance", map["recapture_enterance"].toString())
        }
        val parts: List<MultipartBody.Part> = builder.build().parts()

        launch(UI + QuietCoroutineExceptionHandler) {
            val beanList = LostFoundService.updateReleaseWithPic(lostOrFound, parts).await()

            if (beanList.error_code == -1) {
                this@ReleasePresenterImpl.successCallBack(beanList.data!!)
            }
        }
    }

    override fun uploadReleaseDataWithPic(map: Map<String, Any>, lostOrFound: String, arrayOfFile : ArrayList<File?>) {

        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)

        builder.apply {
            for (i in 0..(arrayOfFile.size - 1)) {
                if (arrayOfFile[i] != null) {
                    val imageBody = RequestBody.create(MediaType.parse("multipart/form-data"), arrayOfFile[i])
                    addFormDataPart("pic[]", arrayOfFile[i]!!.name, imageBody)
                }
            }
            addFormDataPart("title", map["title"].toString())
            addFormDataPart("time", map["time"].toString())
            addFormDataPart("place", map["place"].toString())
            addFormDataPart("detail_type", map["detail_type"].toString())
            addFormDataPart("card_number", map["card_number"].toString())
            addFormDataPart("card_name", map["card_name"].toString())
            addFormDataPart("name", map["name"].toString())
            addFormDataPart("phone", map["phone"].toString())
            addFormDataPart("item_description", map["item_description"].toString())
            addFormDataPart("other_tag", "")
            addFormDataPart("duration", map["duration"].toString())
            addFormDataPart("QQ", map["QQ"].toString())
            addFormDataPart("Wechat", map["Wechat"].toString())
            addFormDataPart("recapture_place", map["recapture_place"].toString())
            addFormDataPart("recapture_enterance", map["recapture_enterance"].toString())
        }
        val parts: List<MultipartBody.Part> = builder.build().parts()

        launch(UI + QuietCoroutineExceptionHandler) {
            val beanList = LostFoundService.updateReleaseWithPic(lostOrFound, parts).await()

            if (beanList.error_code == -1) {
                this@ReleasePresenterImpl.successCallBack(beanList.data!!)
            }
        }
    }

    override fun successCallBack(beanList: List<MyListDataOrSearchBean>) = releaseView.successCallBack(beanList)

    override fun successEditCallback(beanList: List<MyListDataOrSearchBean>) = releaseView.successCallBack(beanList)

    override fun deleteSuccessCallBack(string: String) = releaseView.deleteSuccessCallBack()

    override fun delete(id: Int) {
        launch(UI + QuietCoroutineExceptionHandler) {
            val commonBodyOfId = LostFoundService.delete(id.toString()).await()

            if (commonBodyOfId.error_code == -1) {
                this@ReleasePresenterImpl.deleteSuccessCallBack(commonBodyOfId.data!!)
            }
        }
    }

    override fun uploadEditDataWithPic(map: Map<String, Any>, lostOrFound: String, arrayOfFile: ArrayList<File?>, id: Int) {
        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)

        builder.apply {
            for (i in 0..(arrayOfFile.size - 1)) {
                if (arrayOfFile[i] != null) {
                    val imageBody = RequestBody.create(MediaType.parse("multipart/form-data"), arrayOfFile[i])
                    addFormDataPart("pic[]", arrayOfFile[i]!!.name, imageBody)
                }
            }
            addFormDataPart("title", map["title"].toString())
            addFormDataPart("time", map["time"].toString())
            addFormDataPart("place", map["place"].toString())
            addFormDataPart("detail_type", map["detail_type"].toString())
            addFormDataPart("card_number", map["card_number"].toString())
            addFormDataPart("card_name", map["card_name"].toString())
            addFormDataPart("name", map["name"].toString())
            addFormDataPart("phone", map["phone"].toString())
            addFormDataPart("item_description", map["item_description"].toString())
            addFormDataPart("other_tag", "")
            addFormDataPart("duration", map["duration"].toString())
            addFormDataPart("QQ", map["QQ"].toString())
            addFormDataPart("Wechat", map["Wechat"].toString())
        }
        val parts: List<MultipartBody.Part> = builder.build().parts()
        val anotherLostOrFound = if (Objects.equals(lostOrFound, "editLost")) "lost" else "found"

        launch(UI + QuietCoroutineExceptionHandler) {
            val beanList = LostFoundService.updateEditWithPic(anotherLostOrFound, id.toString(), parts).await()

            if (beanList.error_code == -1) {
                this@ReleasePresenterImpl.successEditCallback(beanList.data!!)
            }
        }
    }

    override fun loadDetailDataForEdit(id: Int, newReleaseView: ReleaseContract.ReleaseView) {
        launch(UI + QuietCoroutineExceptionHandler) {
            val detailData = LostFoundService.getDetailed(id).await()

            if (detailData.error_code == -1) {
                newReleaseView.setEditData(detailData.data!!)
            }
        }
    }
}